### PR TITLE
Add `_foreach_sign`

### PR DESCRIPTION
**Pull request by crcrpar**
Rel:
- #106221

Should we add foreach of [`torch.sgn`](https://pytorch.org/docs/stable/generated/torch.sgn.html) as well?

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -8780,6 +8780,15 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
         supports_autograd=True,
         supports_forward_ad=True,
     ),
+
+    ForeachFuncInfo(
+        'sign',
+        dtypes=floating_types_and(torch.bool, torch.bfloat16, torch.half),
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:8786]: oh interesting that bool is also supported :o

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -8780,6 +8780,15 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
...
+        dtypesIfCUDA=floating_types_and(torch.bfloat16, torch.float16),
+        sample_inputs_func=foreach_inputs_sample_func(1, False, False),
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:8788]: I'm looking in foreach_inputs_sample_func--do you know if we test inplace ops here?

File: torch/optim/rprop.py [hunk 3]
```diff
@@ -307,7 +307,7 @@ def _view_complex_as_real(tensor_list):
                 grouped_prevs[i].copy_(grouped_grads[i])
         grouped_grads = grouped_prevs
 
-        signs = [s.sign() for s in signs]
+        signs = torch._foreach_sign(signs)
```
Comment (janeyx99) [inline torch/optim/rprop.py:310]: beautiful, though could we do
```suggestion
        torch._foreach_sign_(signs)
```

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -10897,6 +10897,21 @@
     CUDA: foreach_tensor_zero_cuda_
   autogen: _foreach_zero, _foreach_zero.out
 
+- func: _foreach_sign(Tensor[] self) -> Tensor[]
```
Comment (janeyx99) [inline aten/src/ATen/native/native_functions.yaml:10900]: This is a nit but I had organized the _foreach_* ops here alphabetically beyond the main add/sub/mul/divs and it would be nice to have this be between sigmoid and sin 😁

**Review by janeyx99 — COMMENTED** [review]
Looks great overall! torch.sgn would but cool to have, but we view complex stuff as real to handle complex so far, so not roadblocking.

A few minor comments.

**Comment by janeyx99** [comment]
This PR would also enable a foreach version of the Lion optim!

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:8786]: I copy paste it from `torch.sign`'s opinfo, IIUC

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:8788]: It should be covered by `test_unary_op`

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:8786]: Yea it makes sense, I was pleasantly intrigued.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:8788]: Ah yes, I see there's _inplace_unary_test now, thank you for the pointer!

**Review by janeyx99 — APPROVED** [review]
Looks great if CI's green!

**Comment by albanD** [comment]
And remove a bunch of for-loops in other optimizers ;)

**Comment by crcrpar** [comment]
enjoyable first try with inductor 🤞
